### ISSUE TITLE
Java 21 support

### ISSUE THREAD

**Issue by minhChuPlaystudios** [mojohaus/aspectj-maven-plugin#186]
Hi guys,
Do you have any plans to add support for java 21 (LTS)?

Thank you in advance!!!

**Comment by slachiewicz** [comment]
Hi, what is missing?

**Comment by minhChuPlaystudios** [comment]
I am using version: 1.14.0
And I got below error

```
[ERROR] Failed to execute goal org.codehaus.mojo:aspectj-maven-plugin:1.14.0:compile (default) on project mongodao: AJC compiler errors:
[ERROR] error source level should be in '1.1'...'1.8','9'...'16' (or '5.0'..'16.0'): 21
[ERROR] error no sources specified
[ERROR] 
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command

```

@slachiewicz : Can you have some ideas for me?

**Comment by slachiewicz** [comment]
Looks like this depends on aspectj update
https://github.com/eclipse-aspectj/aspectj/releases

https://htmlpreview.github.io/?https://github.com/eclipse/org.aspectj/blob/master/docs/dist/doc/README-1920.html

**Comment by minhChuPlaystudios** [comment]
@slachiewicz : Can you tell me when we have release version for this?

**Comment by kriegaex** [comment]
> @slachiewicz : Can you tell me when we have release version for this?

No, he cannot, because the maintainers of this plugin are not participating in AspectJ development. But I have just published a 1.9.21-SNAPSHOT with Java 21 capabilities, which you can try, either with a local build of this plugin of simply with [`dev.aspectj:aspectj-maven-plugin:1.13.1`](https://github.com/dev-aspectj/aspectj-maven-plugin). The latter does not need any updates to be able to run with the latest Java version, you simply set the language level you want. Spoiler: I am the maintainer of both the other plugin and of AspectJ itself.

What keeps me from releasing AspectJ 1.9.21, are some ugly workarounds I need to make it run with the upstream JDT Core (Eclipse Java Compiler, ECJ), which I am still unhappy with. Furthermore, ECJ itself does not support Java 21 preview features yet, only the released ones like record patterns and pattern matching for switch. That would not stop me from releasing, it is just another thing I am unhappy with. But I guess, the snapshot build should get you started. If you think you find any bugs there, please report them to the [AspectJ project](https://github.com/eclipse-aspectj/aspectj/issues).

P.S.: If for any reason your Maven does not find the snapshot repository, you can add this to your POM:

```xml
<repositories>
    <repository>
        <id>ossrh-snapshots</id>
        <url>https://oss.sonatype.org/content/repositories/snapshots</url>
        <releases>
            <enabled>false</enabled>
        </releases>
        <snapshots>
            <enabled>true</enabled>
            <updatePolicy>always</updatePolicy>
        </snapshots>
    </repository>
</repositories>
```

**Comment by rk99d339** [comment]
Hi 
I just tested your 1.9.21-SNAPSHOT with our code. 
There is a problem with Mockito (version 5.6.0) Junit tests. But it only occur when java src level is set to 21.
if we go back to src level 20 it compiles fine. 

The code that makes the error:

`Mockito.when(controller.gIter.hasNext()).thenReturn(true);`

the error Message:
`[ERROR] Cannot invoke thenReturn(boolean) on the primitive type boolean
	/.../.../../../GXXXTest.java:69
when(controller.gIter.hasNext()).thenReturn(true);
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
`

Perhaps you could look into this before release the final version. 
Thanks for your work
Rolf

**Comment by kriegaex** [comment]
@rk99d339, look into what? The text snippets you have just thrown across the fence? Certainly not. If you think the problem is in AspectJ, please provide an [MCVE](https://stackoverflow.com/help/mcve), ideally a GitHub Maven project, and create an issue for AspectJ. Without a minimal reproducer, I will not look into the issue, because I have no intention to speculate about what your code and project configuration exactly look like.

**Comment by rk99d339** [comment]
@kriegaex I can provide you simple test, no problem. 
At the moment we don't have any jdk 21 language features because we couldn't compile them. 
It's a egg and the hen problem.
But we would like to use them in our code in the near future.
I will create you a small application that will reproduce the error.

Sincerely 
Rolf

Sin

**Comment by minhChuPlaystudios** [comment]
@rk99d339 : I have had this problem and just added `Mockito.when()` instead of using only `when()`
I am using mockito-core version 5.5.0

**Comment by rk99d339** [comment]
> @rk99d339 : I have had this problem and just added `Mockito.when()` instead of using only `when()` I am using mockito-core version 5.5.0

That actually helps. But I don't understand why the static import makes a difference. 
I will wait for a fix, because I don't want to go an change all tests cases.. 

Thank you

**Comment by rk99d339** [comment]
@kriegaex 
Here you go. This is an small example to reproduce the bug.
https://github.com/rk99d339/aspectjbug

**Comment by kriegaex** [comment]
Thanks for the reproducer. It still uses the Mojohaus plugin version 1.14.0, which does not even properly support Java 17 yet, not to mention 21. I can reproduce the problem with the AspectJ.dev plugin version 1.13.1, though, because that one supports recent Java versions. I.e., I had to change your sample project first.

But like I said before, **this is not an AspectJ Maven problem,** and you ought to have opened an AspectJ issue for it, so as not to pollute this project here even more.

I found out, it is not even an AspectJ Compiler (AJC) problem as such, but a problem inherited from Eclipse Java Compiler (ECJ). AspectJ depends on it. See https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1518. The good news is that seemingly, the problem was fixed in JDT Core (superset of ECJ), i.e. it should disappear the next time I merge JDT Core into AspectJ. But that step is the most tedious in the whole build process, therefore I only do it when I have enough time, because there are always conflicts to be resolved.

**Comment by kriegaex** [comment]
I merged in the latest ECJ fixes into AJC in https://github.com/eclipse-aspectj/aspectj/issues/263. This fixes the problem for me, but please note that Java 21 support in ECJ is still imperfect, IMO. AspectJ has to wait for the existing issues to be resolved upstream, but at least we are one down.

@rk99d339, please try the latest snapshot I have just deployed.

@slachiewicz, this issue can be closed IMO.

**Comment by slachiewicz** [comment]
Let's keep it open until we have a new release of aspects with initial support for Java21

**Comment by rk99d339** [comment]
Hi. @kriegaex 
Thank you very much!!!
With the newest SNAPSHOT our huge project compiles fine with src level 21

Greetings

**Comment by kriegaex** [comment]
Just in case nobody noticed: Growing tired of waiting for pending JDK 21 fixes in ECJ, I released an [AspectJ milestone 1.9.21.M1](https://repo1.maven.org/maven2/org/aspectj/aspectjtools/1.9.21.M1/) without any release notes. It is available on Maven Central, so you do not have to use a snapshot.
